state.go

```diff
@@ -27,76 +30,848 @@ type State struct {
+	if err == nil {
+		_, err = stream.Read(&signature)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("mls.state: content unmarshal failure %v", err)
+	}
+	_, _ = syntax.Unmarshal(content, &mlsContent)
+
+	pt := &MLSPlaintext{
+		GroupID:           s.GroupID,
+		Epoch:             s.Epoch,
+		Sender:            sender,
+		AuthenticatedData: ct.AuthenticatedData,
+		Content:           mlsContent,
+		Signature:         signature,
+	}
+	return pt, nil
 }
 
-func (s *State) protect(data []byte) *MLSCiphertext {
-	// TODO
-	return nil
+func (s *State) protect(data []byte) (*MLSCiphertext, error) {
+	pt := &MLSPlaintext{
+		GroupID: s.GroupID,
+		Epoch:   s.Epoch,
+		Sender:  s.Index,
+		Content: MLSPlaintextContent{
+			Application: &ApplicationData{
+				Data: data,
+			},
+		},
+	}
+
+	pt.sign(s.groupContext(), s.IdentityPriv, s.scheme)
+	return s.encrypt(pt)
 }
 
 func (s *State) unprotect(ct *MLSCiphertext) ([]byte, error) {
-	// TODO
-	return nil, nil
+	pt, err := s.decrypt(ct)
+	if err != nil {
+		return nil, err
+	}
+
+	sigPubKey := s.Tree.GetCredential(pt.Sender).PublicKey()
+	if !pt.verify(s.groupContext(), sigPubKey, s.scheme) {
+		return nil, fmt.Errorf("invalid message signature")
+	}
+
+	if pt.Content.Type() != ContentTypeApplication {
+		return nil, fmt.Errorf("unprotect attempted on non-application message")
+	}
+	return pt.Content.Application.Data, nil
+}
+
+func senderDataAAD(gid []byte, epoch Epoch, contentType ContentType, nonce []byte) []byte {
+	s := NewWriteStream()
+	err := s.Write(struct {
+		GroupId         []byte `tls:"head=1"`
+		Epoch           Epoch
+		ContentType     ContentType
+		SenderDataNonce []byte `tls:"head=1"`
+	}{
+		GroupId:         gid,
+		Epoch:           epoch,
+		ContentType:     contentType,
+		SenderDataNonce: nonce,
+	})
+
+	if err != nil {
+		return nil
+	}
+
+	return s.Data()
+}
+
+func contentAAD(gid []byte, epoch Epoch,
+	contentType ContentType, authenticatedData []byte,
+	nonce []byte, encSenderData []byte) []byte {
+
+	s := NewWriteStream()
+	err := s.Write(struct {
+		GroupId             []byte `tls:"head=1"`
+		Epoch               Epoch
+		ContentType         ContentType
+		AuthenticatedData   []byte `tls:"head=4"`
+		SenderDataNonce     []byte `tls:"head=1"`
+		EncryptedSenderData []byte `tls:"head=1"`
+	}{
+		GroupId:             gid,
+		Epoch:               epoch,
+		ContentType:         contentType,
+		AuthenticatedData:   authenticatedData,
+		SenderDataNonce:     nonce,
+		EncryptedSenderData: encSenderData,
+	})
+
+	if err != nil {
+		return nil
+	}
+	return s.Data()
+}
+
+func (s State) successor(commit Commit) (*State, error) {
+	// Note: all the slice/map copy operations below on state are mere
+	// reference copies.
+	next := new(State)
+	next.CipherSuite = s.CipherSuite
+	next.GroupID = append(s.GroupID[:0:0], s.GroupID...)
+	next.Epoch = s.Epoch
+	next.Tree = *s.Tree.clone()
+	next.ConfirmedTranscriptHash = append(s.ConfirmedTranscriptHash[:0:0], s.ConfirmedTranscriptHash...)
```

Comment:
Leave this nil, since it should always be updated from the interim hash before use.